lib/Yield/Calls/Slices.cpp

```diff
@@ -0,0 +1,133 @@
+/// \file Slices.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <unordered_map>
+
+#include "llvm/ADT/BreadthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+
+#include "revng/Yield/Calls/Slices.h"
+
+using NodeView = const yield::Graph::Node *;
+
+static yield::Graph::Node *
+nodeHelper(yield::Graph &Graph, const yield::Graph::Node *Source) {
+  auto New = std::make_unique<yield::Graph::Node>(Source->data());
+  return Graph.addNode(std::move(New));
+}
+
+/// \tparam NV local `NodeView` specialization
+/// \tparam INV inverted location `NodeView` specialization
+template<typename NV, typename INV>
+yield::Graph
+preservingSliceImpl(const yield::Graph &Input, const MetaAddress &SlicePoint) {
+  auto LookupHelper = [&SlicePoint](NodeView Node) {
+    return Node->Address == SlicePoint;
+  };
+  auto Entry = llvm::find_if(Input.nodes(), LookupHelper);
+  revng_assert(Entry != Input.nodes().end());
+
+  // Find the rank of each node, such that for any node its rank is equal to
+  // the highest rank among its children plus one.
+  std::unordered_map<const yield::Graph::Node *, size_t> Ranks;
+  std::vector<NodeView> PostOrder(llvm::po_begin(NV{ *Entry }),
+                                  llvm::po_end(NV{ *Entry }));
+  for (NodeView CurrentNode : llvm::reverse(PostOrder)) {
+    uint64_t &CurrentRank = Ranks[CurrentNode];
+
+    CurrentRank = 0;
+    for (auto Child : llvm::children<INV>(CurrentNode))
+      if (auto RankIterator = Ranks.find(Child); RankIterator != Ranks.end())
+        CurrentRank = std::max(RankIterator->second + 1, CurrentRank);
+  }
+
+  // For each node, select a single predecessor to keep connected to.
+  // The ranks calculated earlier are used to choose a specific one.
+  //
+  // TODO: We should consider a better selection algorithm.
+  std::unordered_map<NodeView, NodeView> RealEdges;
+  for (NodeView Node : Input.nodes()) {
+    NodeView SelectedNeighbour = nullptr;
+
+    // This selects a child with the highest rank
+    auto NodeIt = Ranks.find(Node);
+    if (NodeIt != Ranks.end()) {
```

Comment:
So this will always be `!= Ranks.end()`

Reply:
It should, yes, will make this into an assert.

Reply:
Done: https://github.com/revng/revng/pull/191/commits/fd40109649f45f9bf18dbd80692559f30d0300f7.